File: LeetCode/BinaryTree/TreeNode.cs
namespace BinaryTree;

public class TreeNode
{
	public int val;
	public TreeNode? left;
	public TreeNode? right;
	public TreeNode(int val = 0, TreeNode? left=null, TreeNode? right=null)
	{
		this.val = val;
		this.left = left;
		this.right = right;
	}

	public static TreeNode? BuildTree(IReadOnlyList<int?> values) {
		if (values.Count == 0) return null;
        var root = new TreeNode((int)values[0]!);
		var nodes = new Queue<TreeNode?>();
		nodes.Enqueue(root);
		var i = 1;
		while (i<values.Count) {
            var node = nodes.Dequeue();
			if (node is null) {
				nodes.Enqueue(null);
				nodes.Enqueue(null);
				i += 2;
				continue; 
			}
            node.left = values[i] is null ? null : new TreeNode((int)values[i]!);
            nodes.Enqueue(node.left);
			i++;
            node.right = (i < values.Count && values[i] is not null) ? new TreeNode((int)values[i]!) : null;
            nodes.Enqueue(node.right);
            i++;
        }
		return root;
	}

	public static IEnumerable<int?> ToList(TreeNode? root) {
        var result = new List<int?>();
        if (root is null) return result;
		var nodes = new Queue<TreeNode?>();
		nodes.Enqueue(root);
		var nonNullExist = true;
		while (nonNullExist) {
			var nodeCount = nodes.Count;
			nonNullExist = false;
            for (var i=0; i<nodeCount; i++)
			{
                var node = nodes.Dequeue();
                if (node is null)
                {
                    result.Add(null);
                    nodes.Enqueue(null);
                    nodes.Enqueue(null);
                }
                else
                {
	                nonNullExist = true;
	                result.Add(node.val);
	                nodes.Enqueue(node.left);
	                nodes.Enqueue(node.right);
                }
            }
		}

		for (var i = result.Count - 1; i >= 0; i--) {
			if (result[i] is not null) break;
			result.RemoveAt(i);
		}

		return result;
    }
}

File: leetcodePython/TreeNode.py
from collections import deque

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val=val
        self.left = left
        self.right=right

    @staticmethod
    def buildTree(values):
        if not len(values):
            return None
        root = TreeNode(values[0])
        nodes = deque()
        nodes.append(root)
        i=1
        while i < len(values):
            node = nodes.popleft()
            if node is None:
                i += 2
                continue
            if values[i] is None:
                node.left = None
            else:
                node.left = TreeNode(values[i])
            nodes.append(node.left)
            i += 1
            if values[i] is None:
                node.right = None
            else:
                node.right = TreeNode(values[i])
            nodes.append(node.right)
            i += 1
        return root

    @staticmethod
    def toList(root):
        result = []
        if root is None:
            return result
        nodes = deque()
        nodes.append(root)
        nonNullExist = True
        while nonNullExist:
            nodeCount = len(nodes)
            nonNullExist = False
            for i in range(nodeCount):
                node = nodes.popleft()
                if node is None:
                    result.append(None)
                    nodes.append(None)
                    nodes.append(None)
                else:
                    nonNullExist = True
                    result.append(node.val)
                    nodes.append(node.left)
                    nodes.append(node.right)
        for t in reversed(range(len(result))):
            if result[t] is not None:
                break
            result.pop(t)

        return result